chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,407 @@
+
+     ;;;;; end library code
+
+     ;; return the puzzle hash for a cat with the given `GENESIS_COIN_CHECKER_hash` & `INNER_PUZZLE`
+     (defun-inline cat_puzzle_hash (cat_mod_struct inner_puzzle_hash)
+       (puzzle-hash-of-curried-function (mod_hash_from_cat_mod_struct cat_mod_struct)
+                                        inner_puzzle_hash
+                                        (sha256 ONE (genesis_coin_checker_hash_from_cat_mod_struct cat_mod_struct))
+                                        (mod_hash_hash_from_cat_mod_struct cat_mod_struct)
+       )
+     )
+
+     ;; tweak `CREATE_COIN` condition by wrapping the puzzle hash, forcing it to be a cat
+     ;; prohibit CREATE_COIN_ANNOUNCEMENT
+
+     (defun-inline morph_condition (condition cat_mod_struct)
+       (if (= (f condition) CREATE_COIN)
+         (c CREATE_COIN
+               (c (cat_puzzle_hash cat_mod_struct (f (r condition)))
+                    (r (r condition)))
+         )
+         (if (= (f condition) CREATE_COIN_ANNOUNCEMENT)
+           (list CREATE_COIN_ANNOUNCEMENT (concat 0xca (f (r condition))))
+           condition
+         )
+       )
+     )
+
+     ;; tweak all `CREATE_COIN` conditions, enforcing created coins to be cats
+     (defun morph_conditions (conditions cat_mod_struct)
+       (if conditions
+         (c
+           (morph_condition (f conditions) cat_mod_struct)
+           (morph_conditions (r conditions) cat_mod_struct)
+         )
+         ()
+       )
+     )
+
+     ;; given a coin triplet, return the id of the coin
+     (defun-inline coin_id_for_coin (coin cat_mod_struct)
+       (sha256 (f coin) (cat_puzzle_hash cat_mod_struct (f (r coin))) (f (r (r coin))))
+     )
+
+     ;; utility to fetch coin amount from coin
+     (defun-inline input_amount_for_coin (coin)
+       (f (r (r coin)))
+     )
+
+     ;; calculate the hash of an announcement
+     (defun-inline calculate_annoucement_id (this_coin_id this_subtotal next_coin_id cat_mod_struct)
+       ; NOTE: the next line containts a bug, as sha256tree1 ignores `this_subtotal`
+       (sha256 next_coin_id 0xcb this_coin_id this_subtotal)
+     )
+
+     ;; create the `ASSERT_COIN_ANNOUNCEMENT` condition that ensures the next coin's announcement is correct
+     (defun-inline create_assert_next_announcement_condition (this_coin_id this_subtotal next_coin_id cat_mod_struct)
+       (list ASSERT_COIN_ANNOUNCEMENT
+             (calculate_annoucement_id this_coin_id
+                                            this_subtotal
+                                            next_coin_id
+                                            cat_mod_struct
+             )
+       )
+     )
+
+     ;; here we commit to I_{k-1} and S_k
+     (defun-inline create_announcement_condition (prev_coin_id prev_subtotal)
+       (list CREATE_COIN_ANNOUNCEMENT
+             (concat 0xcb (concat prev_coin_id prev_subtotal))
```

Comment:
If you change the `concat 0xcb` above to `sha256`, then you have to do the same here too.

Reply:
(And change some python solution-generating code too.)

Reply:
At the very least, get rid of one of the `concat` operators here.
```suggestion
             (concat 0xcb prev_coin_id prev_subtotal)
```

I think that
```suggestion
             (sha256 0xcb prev_coin_id prev_subtotal)
```

is better though (although other things in python would have to change in this case too).

I very much hate `concat`. `sha256` is almost always a better option.